include/revng/Pipeline/InspectableLLVMContainer.h

```diff
@@ -0,0 +1,171 @@
+
+  virtual TargetsList
+  compactTargets(const Context &Ctx, TargetsList::List &Targets) const = 0;
+
+  llvm::DenseSet<const llvm::Function *>
+  enumerateFunctions(const TargetsList &Targets,
+                     const LLVMContainer &Container) const {
+
+    llvm::DenseSet<const llvm::Function *> ToReturn;
+    for (auto &GL : Container.getModule().functions()) {
+      auto MaybeTarget = symbolToTarget(GL);
+      if (not MaybeTarget.has_value())
+        continue;
+
+      if (not Targets.contains(*MaybeTarget))
+        continue;
+
+      ToReturn.insert(&GL);
+    }
+
+    return ToReturn;
+  }
+
+  llvm::DenseSet<llvm::Function *>
+  enumerateFunctions(const TargetsList &Targets,
+                     LLVMContainer &Container) const {
+    llvm::DenseSet<llvm::Function *> ToReturn;
+    for (auto &GL : Container.getModule().functions()) {
+      auto MaybeTarget = symbolToTarget(GL);
+      if (not MaybeTarget.has_value())
+        continue;
+
+      if (not Targets.contains(*MaybeTarget))
+        continue;
+
+      ToReturn.insert(&GL);
+    }
+
+    return ToReturn;
+  }
+
+  bool remove(const Context &Ctx,
+              const TargetsList &Targets,
+              LLVMContainer &Container) const final {
+
+    const bool AllContained = enumerate(Ctx, Container).contains(Targets);
+
+    for (auto &GL : enumerateFunctions(Targets, Container))
+      GL->deleteBody();
+
+    return AllContained;
+  }
+
+  TargetsList
+  enumerate(const Context &Ctx, const LLVMContainer &Container) const final {
+    TargetsList::List L;
+    for (auto &GL : Container.getModule().functions()) {
+      auto MaybeTarget = symbolToTarget(GL);
+      if (not MaybeTarget.has_value())
+        continue;
+
+      L.push_back(std::move(*MaybeTarget));
+    }
+
+    return compactTargets(Ctx, L);
+  }
+
+  static llvm::DenseSet<const llvm::Function *>
+  enumerateNonOnwedFunctions(const LLVMContainer &Container) {
+    llvm::DenseSet<const llvm::Function *> ToReturn;
+
+    for (const auto &F : Container.getModule().functions())
+      if (not hasOwner(F))
+        ToReturn.insert(&F);
+
+    return ToReturn;
+  }
+
+  static llvm::DenseSet<const llvm::Function *>
+  enumerateAllFunctions(const TargetsList &Targets,
+                        const LLVMContainer &Container) {
+    llvm::DenseSet<const llvm::Function *> ToReturn;
+
+    for (const auto &Kind : getRegisteredInspectors()) {
+      auto Set = Kind->enumerateFunctions(Targets, Container);
+      for (const auto &E : Set)
+        ToReturn.insert(E);
+    }
+
+    return ToReturn;
+  }
+
+private:
+  static StaticContainer &getRegisteredInspectors() {
+    static StaticContainer Container;
+    return Container;
+  }
+};
+
+/// Convienency class that can be used to declare a kind and to specify
+/// how find their instances into a llvm container at the same time
+template<typename LLVMContainer>
+class KindInspector : public Kind,
+                      public LLVMContainerInspector<LLVMContainer> {
```

Comment:
* [ ] `KindInspector` is LLVM-specific but the name doesn't show it.
  `LLVMKindInspector` or maybe just `LLVMKind`?
* [ ] Do we really need a template argument here? `LLVMContainer` is used everywhere with the actual class `LLVMContainer`. Same for `LLVMContainerInspector`. I understand that it's a degree of freedom one day we might want to use, but it seems a bit arbitrary to introduce it now.
  My view is: there is a revng-pipeline library that has some degree of flexibility for other unforeseen use cases, but I think LLVM is a feature that's specific to revng and that doesn't need to be super-mega-generic.

Reply:
i will close this one since there is another one saying the same thing